blog/gu-kelly-xiu-replication/index.qmd

```diff
@@ -0,0 +1,475 @@
+
+We rely on the preparation of the macroeconomic predictors from the corresponding chapter in [Accessing and Managing Financial Data](https://www.tidy-finance.org/r/accessing-and-managing-financial-data.html#macroeconomic-predictors). The following code snippets merges the stock-level characteristics with the CRSP sample and the macroeconomic predictors. Note that no further adjustment of the timing of excess returns and stock-level characteristics is necessary:
+
+> In this dataset, we've already adjusted the lags. (e.g. When DATE=19570329 in our dataset, you can use the monthly RET at 195703 as the response variable.) (readme.txt from the available dataset on Dacheng Xiu's homepage). 
+
+To remain consistent, we only lag the macropredictors by one month.  
+
+```{r}
+tidy_finance <- dbConnect(SQLite(), 
+                          "data/tidy_finance_r.sqlite",
+                          extended_types = TRUE)
+
+crsp_monthly <- tbl(tidy_finance, "crsp_monthly") |>
+  select(month, permno, mktcap_lag, ret_excess) |>
+  collect()
+
+macro_predictors <- tbl(tidy_finance, "macro_predictors") |>
+  select(month, dp, ep, bm, ntis, tbl, tms, dfy, svar) |>
+  collect() |>
+  rename_with(~ paste0("macro_", .), -month)
+
+macro_predictors <- macro_predictors |>
+  select(month) |>
+  left_join(macro_predictors |> mutate(month = month %m+% months(1)), 
+            join_by(month))
+
+characteristics <- characteristics |>
+  inner_join(crsp_monthly, join_by(month, permno)) |>
+  inner_join(macro_predictors, join_by(month)) |>
+  arrange(month, permno) |>
+  mutate(macro_intercept = 1) |>
+  select(permno, month, ret_excess, mktcap_lag, 
+         sic2, contains("macro"), contains("characteristic"))
+```
+
+### Create dataset with all interaction terms
+
+Before applying machine-learning models, the authors include interaction terms between macroeconomic predictors and stock characteristics in the sample. 
+
+> We include 74 industry dummies corresponding to the first two digits of Standard Industrial Classification (SIC) codes. [...] [The stock-level covariates also include] interactions between stock-level characteristics and macroeconomic state variables. The total number of covariates is 94×(8+1)+74=920. (p. 2249)
+
+The following recipe creates dummies for each `sic2` code and adds interaction terms between macro variables and stock characteristics. Note that `data` for the recipe is only required to identify the columns in the sample. No computations are performed at the first step. For that reason, I call `head()` to keep the computational burden a bit lower without affecting the outcome. 
+
+```{r}
+rec <- recipe(ret_excess ~ ., data = characteristics |> head())|>
+  update_role(permno, month, mktcap_lag, 
+              new_role = "id") |>
+  step_interact(terms = ~ contains("characteristic"):contains("macro"), 
+                keep_original_cols = FALSE) |>
+  step_dummy(sic2, 
+             one_hot = TRUE)
+
+rec <- rec |> prep()
```

Comment:
In these cases, I'd just write prep(rec). Piping makes imho little sense for one step function calls with single inputs.